subsys/instrumentation/common/instr_common.c

```diff
@@ -0,0 +1,528 @@
+/*
+ * Copyright 2023 Linaro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/instrumentation/instrumentation.h>
+#include <zephyr/instrumentation/instrumentation_ringbuffer.h>
+#include <zephyr/instrumentation/instrumentation_timestamp.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/kernel.h>
+#include <zephyr/kernel_structs.h>
+#include <zephyr/retention/retention.h>
+#include <zephyr/sys/reboot.h>
+
+#include <kernel_internal.h>
+#include <ksched.h>
+
+/*
+ * Memory buffer to store instrumentation event records has the following modes:
+ *
+ * Callgraph ring buffer (default): Replace oldest entry when buffer full, at
+ * the cost of processing stray entry events that have been truncated.
+ *
+ * Callgraph fixed buffer: Stop buffering events when the buffer is full,
+ * ensuring we have a callgraph from reset or from whenever the trigger event
+ * was entered.
+ *
+ * Statistical (profiling): Buffer functions until out of memory.
+ *
+ */
+
+const struct device *retention0 = DEVICE_DT_GET(DT_NODELABEL(retention0));
```

Comment:
I'm probably missing some context here but why do we want to use some retained memory here?

 What about if I simply want to use some memory area (RAM) to store the data? Maybe I have some SHMEM that I want to use that is accessible from some other core.

Reply:
The retained mem size used is only 32 bytes, to keep the `trigger` and `stopper` addresses between reboots, as I explained above. The case you mention here - that we've discussed in the ZDS -- is super cool and I think it should be just another backend, which would be accessed via the `zaru` CLI tool. We can decouple the retention subsys, but then zaru CLI won't work anymore. I like the CLI tool because you can change the region to be  traced at runtime, easily, without going to set it in `Kconfig`. (You can set defaults at compile time for the `trigger` and `stopper` functions in Kconfig`).

Reply:
Retention works in uninitialised RAM already, if it needs additional code to work across another core than it can be added as a new retention driver, that's not a problem

Reply:
@gromero couldn't the trigger and stopper address be stored and restored using the zaru CLI tool. If that is not possible, why not just use a memory section and avoid all retained mem overhead.